### ISSUE TITLE
EntityManagerContext is not destroyed after commit of transaction started with @Transactional on type

### ISSUE THREAD

**Issue by andreysubbotin** [cuba-platform/cuba#2677]
https://youtrack.cuba-platform.com/issue/PL-6653

### Environment

- Platform version: 7.1

### Description of the bug or enhancement

- Minimal reproducible example
Entity modified in some service annotated with @Transactional (annotation on class), so entity log entry created.
Then another transaction starts (e.g. reload). The system tries to insert EntityLogItem object that is already committed to the database.
```
2016-01-25 17:54:05,351 ERROR [com.haulmont.cuba.security.app.EntityLog] [app-core] [admin] Error saving EntityLogItem
java.sql.SQLException: ERROR: duplicate key value violates unique constraint "sec_entity_log_pkey"
  Подробности: Key (id)=(e2903b69-e4f1-875e-0fca-ab36d4afa9df) already exists. {prepstmnt 1424223722 
insert into SEC_ENTITY_LOG 
        (ID,CREATE_TS,CREATED_BY,EVENT_TS,USER_ID,CHANGE_TYPE,ENTITY,ENTITY_ID,CHANGES) 
    VALUES (?,current_timestamp,?,?,?,?,?,?,?) 
[params=(Object) e2903b69-e4f1-875e-0fca-ab36d4afa9df, (Object) admin, (Object) 2016-01-25 17:54:05.326, (Object) 60885987-1b61-4247-94c7-dff348347f93, (Object) M, (Object) ref$Car, (Object) 20318b46-ea01-30bb-5f1e-fdf4ece85f14, (Object) vin=0.46286084695230456
]} [code=0, state=23505] Query: insert into SEC_ENTITY_LOG (ID,CREATE_TS,CREATED_BY,EVENT_TS,USER_ID,CHANGE_TYPE,ENTITY,ENTITY_ID,CHANGES) values (?,current_timestamp,?,?,?,?,?,?,?) Parameters: [e2903b69-e4f1-875e-0fca-ab36d4afa9df, admin, 2016-01-25 17:54:05.326, 60885987-1b61-4247-94c7-dff348347f93, M, ref$Car, 20318b46-ea01-30bb-5f1e-fdf4ece85f14, vin=0.46286084695230456
]
    at com.haulmont.bali.db.QueryRunner.rethrow(QueryRunner.java:453)
    at com.haulmont.bali.db.QueryRunner.update(QueryRunner.java:541)
    at com.haulmont.bali.db.QueryRunner.update(QueryRunner.java:512)
    at com.haulmont.cuba.security.app.EntityLog.saveItem(EntityLog.java:99)
    at com.haulmont.cuba.security.app.EntityLog.flush(EntityLog.java:88)
    at com.haulmont.cuba.core.sys.persistence.EntityTransactionListener.afterCommit(EntityTransactionListener.java:59)
    at org.apache.openjpa.event.TransactionEventManager.fireEvent(TransactionEventManager.java:107)
    at org.apache.openjpa.lib.util.concurrent.AbstractConcurrentEventManager.fireEvent(AbstractConcurrentEventManager.java:115)
    at org.apache.openjpa.kernel.BrokerImpl.fireTransactionEvent(BrokerImpl.java:896)
    at org.apache.openjpa.kernel.BrokerImpl.endTransaction(BrokerImpl.java:2407)
    at org.apache.openjpa.kernel.BrokerImpl.afterCompletion(BrokerImpl.java:2043)
    at org.apache.openjpa.kernel.LocalManagedRuntime.commit(LocalManagedRuntime.java:94)
    at org.apache.openjpa.kernel.BrokerImpl.commit(BrokerImpl.java:1528)
    at org.apache.openjpa.kernel.DelegatingBroker.commit(DelegatingBroker.java:933)
    at org.apache.openjpa.persistence.EntityManagerImpl.commit(EntityManagerImpl.java:570)
    at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:514)
    at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:755)
    at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:724)
    at com.haulmont.cuba.core.sys.TransactionImpl.commit(TransactionImpl.java:65)
    at com.haulmont.cuba.core.app.DataManagerBean.load(DataManagerBean.java:112)
    at com.haulmont.cuba.core.app.DataServiceBean.load(DataServiceBean.java:36)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
    at org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:80)
    at com.haulmont.cuba.core.sys.ServiceInterceptor.aroundInvoke(ServiceInterceptor.java:64)
    at sun.reflect.GeneratedMethodAccessor93.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:621)
    at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:610)
    at org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:65)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:161)
    at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:91)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
    at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
    at com.sun.proxy.$Proxy171.load(Unknown Source)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at com.haulmont.cuba.core.sys.remoting.LocalServiceInvokerImpl.invoke(LocalServiceInvokerImpl.java:71)
    at com.haulmont.cuba.web.sys.remoting.LocalServiceProxy$LocalServiceInvocationHandler.invoke(LocalServiceProxy.java:115)
    at com.sun.proxy.$Proxy16.load(Unknown Source)
    at com.haulmont.cuba.client.sys.DataManagerClientImpl.load(DataManagerClientImpl.java:38)
    at com.haulmont.cuba.client.sys.DataManagerClientImpl.reload(DataManagerClientImpl.java:79)
    at com.haulmont.cuba.client.sys.DataManagerClientImpl.reload(DataManagerClientImpl.java:65)
    at com.haulmont.cuba.client.sys.DataManagerClientImpl.reload(DataManagerClientImpl.java:60)
    at com.haulmont.cuba.client.sys.DataManagerClientImpl.reload(DataManagerClientImpl.java:55)
    at com.haulmont.cuba.gui.data.impl.GenericDataSupplier.reload(GenericDataSupplier.java:40)
    at com.haulmont.refapp.gui.app.car.CarBrowser$2.actionPerform(CarBrowser.java:82)
    at com.haulmont.cuba.web.gui.components.WebButton.performAction(WebButton.java:43)
    at com.haulmont.cuba.web.gui.components.WebButton$1.buttonClick(WebButton.java:34)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at com.vaadin.event.ListenerMethod.receiveEvent(ListenerMethod.java:508)
    at com.vaadin.event.EventRouter.fireEvent(EventRouter.java:198)
    at com.vaadin.event.EventRouter.fireEvent(EventRouter.java:161)
    at com.vaadin.server.AbstractClientConnector.fireEvent(AbstractClientConnector.java:986)
    at com.vaadin.ui.Button.fireClick(Button.java:393)
    at com.haulmont.cuba.web.toolkit.ui.CubaButton.fireClick(CubaButton.java:42)
    at com.vaadin.ui.Button$1.click(Button.java:61)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at com.vaadin.server.ServerRpcManager.applyInvocation(ServerRpcManager.java:168)
    at com.vaadin.server.ServerRpcManager.applyInvocation(ServerRpcManager.java:118)
    at com.vaadin.server.communication.ServerRpcHandler.handleInvocations(ServerRpcHandler.java:291)
    at com.vaadin.server.communication.ServerRpcHandler.handleRpc(ServerRpcHandler.java:184)
    at com.vaadin.server.communication.UidlRequestHandler.synchronizedHandleRequest(UidlRequestHandler.java:92)
    at com.vaadin.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:41)
    at com.vaadin.server.VaadinService.handleRequest(VaadinService.java:1408)
    at com.vaadin.server.VaadinServlet.service(VaadinServlet.java:355)
    at com.haulmont.cuba.web.sys.CubaApplicationServlet.serviceAppRequest(CubaApplicationServlet.java:210)
    at com.haulmont.cuba.web.sys.CubaApplicationServlet.service(CubaApplicationServlet.java:135)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:731)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
    at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
    at com.haulmont.cuba.web.sys.CubaHttpFilter.handleNotFiltered(CubaHttpFilter.java:99)
    at com.haulmont.cuba.web.sys.CubaHttpFilter.doFilter(CubaHttpFilter.java:86)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
    at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
    at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
    at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:505)
    at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:170)
    at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
    at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:957)
    at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
    at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:423)
    at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1079)
    at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:620)
    at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:316)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
    at java.lang.Thread.run(Thread.java:745)
2016-01-25 17:54:05,374 ERROR [com.haulmont.cuba.security.app.EntityLog] [app-core] [admin] Error saving EntityLogItem
java.sql.SQLException: ERROR: duplicate key value violates unique constraint "sec_entity_log_pkey"
  Подробности: Key (id)=(8dfbdc6e-4d97-0e50-d76a-e8c34e69feca) already exists. {prepstmnt 539639004 
insert into SEC_ENTITY_LOG 
        (ID,CREATE_TS,CREATED_BY,EVENT_TS,USER_ID,CHANGE_TYPE,ENTITY,ENTITY_ID,CHANGES) 
    VALUES (?,current_timestamp,?,?,?,?,?,?,?) 
[params=(Object) 8dfbdc6e-4d97-0e50-d76a-e8c34e69feca, (Object) admin, (Object) 2016-01-25 17:54:05.325, (Object) 60885987-1b61-4247-94c7-dff348347f93, (Object) M, (Object) ref$Car, (Object) 1d6e981a-3c48-443d-1396-b4c96520357d, (Object) vin=0.29002673922751954
]} [code=0, state=23505] Query: insert into SEC_ENTITY_LOG (ID,CREATE_TS,CREATED_BY,EVENT_TS,USER_ID,CHANGE_TYPE,ENTITY,ENTITY_ID,CHANGES) values (?,current_timestamp,?,?,?,?,?,?,?) Parameters: [8dfbdc6e-4d97-0e50-d76a-e8c34e69feca, admin, 2016-01-25 17:54:05.325, 60885987-1b61-4247-94c7-dff348347f93, M, ref$Car, 1d6e981a-3c48-443d-1396-b4c96520357d, vin=0.29002673922751954
]
    at com.haulmont.bali.db.QueryRunner.rethrow(QueryRunner.java:453)
    at com.haulmont.bali.db.QueryRunner.update(QueryRunner.java:541)
    at com.haulmont.bali.db.QueryRunner.update(QueryRunner.java:512)
    at com.haulmont.cuba.security.app.EntityLog.saveItem(EntityLog.java:99)
    at com.haulmont.cuba.security.app.EntityLog.flush(EntityLog.java:88)
    at com.haulmont.cuba.core.sys.persistence.EntityTransactionListener.afterCommit(EntityTransactionListener.java:59)
    at org.apache.openjpa.event.TransactionEventManager.fireEvent(TransactionEventManager.java:107)
    at org.apache.openjpa.lib.util.concurrent.AbstractConcurrentEventManager.fireEvent(AbstractConcurrentEventManager.java:115)
    at org.apache.openjpa.kernel.BrokerImpl.fireTransactionEvent(BrokerImpl.java:896)
    at org.apache.openjpa.kernel.BrokerImpl.endTransaction(BrokerImpl.java:2407)
    at org.apache.openjpa.kernel.BrokerImpl.afterCompletion(BrokerImpl.java:2043)
    at org.apache.openjpa.kernel.LocalManagedRuntime.commit(LocalManagedRuntime.java:94)
    at org.apache.openjpa.kernel.BrokerImpl.commit(BrokerImpl.java:1528)
    at org.apache.openjpa.kernel.DelegatingBroker.commit(DelegatingBroker.java:933)
    at org.apache.openjpa.persistence.EntityManagerImpl.commit(EntityManagerImpl.java:570)
    at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:514)
    at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:755)
    at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:724)
    at com.haulmont.cuba.core.sys.TransactionImpl.commit(TransactionImpl.java:65)
    at com.haulmont.cuba.core.app.DataManagerBean.load(DataManagerBean.java:112)
    at com.haulmont.cuba.core.app.DataServiceBean.load(DataServiceBean.java:36)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
    at org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:80)
    at com.haulmont.cuba.core.sys.ServiceInterceptor.aroundInvoke(ServiceInterceptor.java:64)
    at sun.reflect.GeneratedMethodAccessor93.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:621)
    at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:610)
    at org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:65)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:161)
    at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:91)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
    at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
    at com.sun.proxy.$Proxy171.load(Unknown Source)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at com.haulmont.cuba.core.sys.remoting.LocalServiceInvokerImpl.invoke(LocalServiceInvokerImpl.java:71)
    at com.haulmont.cuba.web.sys.remoting.LocalServiceProxy$LocalServiceInvocationHandler.invoke(LocalServiceProxy.java:115)
    at com.sun.proxy.$Proxy16.load(Unknown Source)
    at com.haulmont.cuba.client.sys.DataManagerClientImpl.load(DataManagerClientImpl.java:38)
    at com.haulmont.cuba.client.sys.DataManagerClientImpl.reload(DataManagerClientImpl.java:79)
    at com.haulmont.cuba.client.sys.DataManagerClientImpl.reload(DataManagerClientImpl.java:65)
    at com.haulmont.cuba.client.sys.DataManagerClientImpl.reload(DataManagerClientImpl.java:60)
    at com.haulmont.cuba.client.sys.DataManagerClientImpl.reload(DataManagerClientImpl.java:55)
    at com.haulmont.cuba.gui.data.impl.GenericDataSupplier.reload(GenericDataSupplier.java:40)
    at com.haulmont.refapp.gui.app.car.CarBrowser$2.actionPerform(CarBrowser.java:82)
    at com.haulmont.cuba.web.gui.components.WebButton.performAction(WebButton.java:43)
    at com.haulmont.cuba.web.gui.components.WebButton$1.buttonClick(WebButton.java:34)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at com.vaadin.event.ListenerMethod.receiveEvent(ListenerMethod.java:508)
    at com.vaadin.event.EventRouter.fireEvent(EventRouter.java:198)
    at com.vaadin.event.EventRouter.fireEvent(EventRouter.java:161)
    at com.vaadin.server.AbstractClientConnector.fireEvent(AbstractClientConnector.java:986)
    at com.vaadin.ui.Button.fireClick(Button.java:393)
    at com.haulmont.cuba.web.toolkit.ui.CubaButton.fireClick(CubaButton.java:42)
    at com.vaadin.ui.Button$1.click(Button.java:61)
```

**Comment by andreabat** [comment]
Hi, i think this occurs on version 6.10.x as well. Is it going to be fixed there as well ?

**Comment by andreysubbotin** [comment]
Unfortunately, we have no plans to fix it on 6.10. As a workaround, you can place `@Transactional` directly on methods. This should work.
